### ISSUE TITLE
Aria-required-children violation due to `input` within `grid`

### ISSUE THREAD

**Issue by clamli** [dequelabs/axe-core#4162]
### Product

axe-core

### Question

[Angular material has an example that uses an input in a `mat-chip-grid`](https://material.angular.io/components/chips/overview#text-entry).

Using an input with the `role="combobox"` will trigger an `aria-required-children` violation. However, this is a common pattern for filter input and has been widely used. The following is a simplified DOM that contains the violation,

```html
<mat-form-field>
  <mat-chip-grid #myChipGrid [(ngModel)]="mySelection"
  aria-label="enter sandwich fillings">
    <mat-chip-row *ngFor="let filling of fillings"
                 (removed)="remove(filling)">
      {{filling.name}}
      <button matChipRemove>
        <mat-icon>cancel</mat-icon>
      </button>
    </mat-chip-row>
    <input type="text"
           aria-label="Add a chip"
           role="combobox"
           aria-expanded="false"
           aria-haspopup="listbox"
           id="chip-input"
           [matChipInputFor]="myChipGrid"
           [matChipInputSeparatorKeyCodes]="separatorKeysCodes"
           (matChipInputTokenEnd)="add($event)"/>
  </mat-chip-grid>
</mat-form-field>
```

1. What is your suggestion on fixing this violation? Should we address this by moving the input outside of the grid? 
2. Since an input with `role="combobox" does not appear to generate any accessibility issues, has there been any validation that this check is not a false positive?

**Comment by straker** [comment]
Thanks for the issue. Could you tell me which version of axe-core you are using and more about what the `aria-required-children` violation is pointing to? I ask as since [axe-core v4.3.0](https://github.com/dequelabs/axe-core/releases/tag/v4.3.0) the combobox role no longer requires a required child. Using latest axe-core v4.8.1 and I'm not seeing any violations for your provided example.

**Comment by clamli** [comment]
Hi @straker, 

We are using axe-core 4.7.1 right now.

**Failure Summary:** Fix any of the following: Element has children which are not allowed: [role=combobox] Element uses aria-busy="true" while showing a loader

**Example being tested:** https://stackblitz.com/edit/gvaswn?file=src%2Fexample%2Fchips-autocomplete-example.html

> I ask as since [axe-core v4.3.0](https://github.com/dequelabs/axe-core/releases/tag/v4.3.0) the combobox role no longer requires a required child

Does it mean a combobox role under a parent whose required children do not contain combobox will not trigger `aria-required-children` violation?

**Comment by clamli** [comment]
Hi @straker,

Thanks for your explanation. I now understand that this example will trigger an `aria-required-children` violation because the parent element contains a node that is not classified as one of its direct children. 

It is my understanding that this was not previously considered an error, as long as the parent element contained a child element with the required role, `role=row` in this case. However, the new rule is more strict and does not allow any disallowed child elements.

I am curious about the reasoning behind that change. And in the case of the `mat-chip-grid` and `matChipInput` example above, since it is already a widely used pattern to place an input element inside a `mat-chip-grid` (Please see the example on the Angular Material page) and does not cause any accessibility issues other than triggering the violation, would we consider it a valid false positive report?

**Comment by straker** [comment]
I'm trying to remember the exact reason but I'm afraid I can't recall them. @WilcoFiers should be able to answer. I'll let him know to come look at this next week.

**Comment by clamli** [comment]
Hi @WilcoFiers! 

Kindly ask for an update.

**Comment by straker** [comment]
@clamli Sorry, will probably be another week or so as Wilco is unavailable to respond at the moment.

**Comment by WilcoFiers** [comment]
/me waves. Back from vacation.

The reason for this change is simply that it creates problems to put unexpected elements inside grouping roles like list, grid, tree, etc. Based case scenario it miscounts the number of items in the grouping. Announcing 3 chips when there are only two. That can throw people off. Worse though would be that the AT tries to compensate and decides to ignore the unexpected element.

VoiceOver seems to do that with this `mat-clip-grid` component. When it reaches this grid, it switches to table mode. In table mode it ignores anything that's not inside a cell (i.e., the input element), and it stays in table mode until it the grid is closed. Effectively skipping that input element. There are ways to prompt VoiceOver to go directly to that input element if a user realises they missed something, but they might not.


ARIA 1.3 is introducing the concept of [Allowed accessible child roles](https://w3c.github.io/aria/#mustContain), which is making this a more explicit requirement of ARIA now, but it has always been an issue, and can break in unexpected ways because assistive tech tries to compensate for errors in different ways.
